### ISSUE TITLE
Admin Approval / Admin Consent required (Microsoft Azure AD Error AADSTS90094)

### ISSUE THREAD

**Issue by webe3** [abraunegg/onedrive#769]
I have been using OneDrive free client happily for a while.  The company I work for recently was bought out.  As part of that, I was migrated to the parent company's OneDrive instance.  When I go through the setup process, now I get a screen that says administrator approval is required for OneDrive free client.  I never had to do this before with the previous company.  I do not have administrative rights so I have had to reach out to the help desk.  They are of course outsourced to the cheapest country of the month and do not have anything like this in their scripts.  Can you tell me exactly what is required of the administrator and where it is done so that I can lead them through the process.  That is if they are even willing to do it.  I will probably get some type of security push back.  Any tips on how to convince them it is not a security risk would also be appreciated.  Obviously the previous company OneDrive setup allowed access from the onedrive free client because it was working.  I doubt if the help desk has enough going for them to investigate how the old onedrive was set up to see what needs to be done with the new one.  As much information as you can give me would be appreciated as I will have to lead the help desk by the hand blindly as I do not have admin rights to even see for myself what needs to be done.  I found this closed issue but it doesn't give me the details I need "OneDrive for business - Admin Consent #16".  Since I am not the first to have to go through this, maybe even some help write up in the setup instructions would be useful.

**Comment by abraunegg** [comment]
@webe3 

This is not something that I can assist with as this is an issue with your Azure Active Directory configuration.

> Can you tell me exactly what is required of the administrator and where it is done so that I can lead them through the process

Have a read of these URL's: 
https://docs.microsoft.com/en-us/archive/blogs/aaddevsup/receiving-aadsts90094-the-grant-requires-admin-permission

https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent

https://stackoverflow.com/questions/44081476/azure-ad-admin-consent-required-when-it-shouldnt

> I doubt if the help desk has enough going for them to investigate how the old onedrive was set up to see what needs to be done with the new one. As much information as you can give me would be appreciated as I will have to lead the help desk by the hand blindly as I do not have admin rights to even see for myself what needs to be done

Suggest that you find out who (not helpdesk) in your new company is responsible for Azure Active Directory. Work with them directly to solve this by either:

* Tweaking the Azure AD configuration
* Generating an Admin Consent Auth URL that they use to login with, which will authenticate the application for all users

**Comment by abraunegg** [comment]
@webe3 
Also, as per #16, I have no idea how @skilion configured the Microsoft Azure App registration - no access, no visibility.

When I create a new app registration however, below is what I am presented with:

**Graph API Permissions**
![azure_app_1](https://user-images.githubusercontent.com/4956234/71922544-3aab6300-31df-11ea-9010-2f5f608fbb72.jpg)

**Permissions for Files** 
![azure_app_2](https://user-images.githubusercontent.com/4956234/71922578-4d259c80-31df-11ea-9ed7-9a0d4e3f3268.jpg)

**Permissions for Sites**
![azure_app_3](https://user-images.githubusercontent.com/4956234/71922592-56af0480-31df-11ea-8525-4021160846b3.jpg)

**Overall Permissions**
![azure_app_4](https://user-images.githubusercontent.com/4956234/71922612-60d10300-31df-11ea-82cc-7e952e82d0dd.jpg)

None of these appear to require 'Admin Consent', however if I select to use "Application permissions", then 'Admin Consent' is required.

![azure_app_5](https://user-images.githubusercontent.com/4956234/71922847-de950e80-31df-11ea-98b7-bc8910ca7426.jpg)

Without knowing what was originally configured, potentially 'both' was selected, thus potentially leading to this issue.

What I can do is potentially see if this new app registration ID works in my testing, and if it does, generate a PR and see if that works for you.

**Comment by abraunegg** [comment]
@webe3 
I have tested using an updated client id which has the above permissions configured. Client was able to authenticate, upload / download files.

Please can you test this PR against your issue and provide feedback:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/770/head:pr770
git checkout pr770
./configure; make clean; make;
```

**Comment by webe3** [comment]
I am running on Debian and had to install a few packages to try to get the build to work.  It is errorring out on curl but I do have it installed.
./configure; make clean; make;
checking for a BSD-compatible install... /usr/bin/install -c
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for dmd... no
checking for ldmd2... ldmd2
checking version of D compiler... 1.18.0
checking for curl... no
configure: error: Package requirements (libcurl) were not met:

No package 'libcurl' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables curl_CFLAGS
and curl_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
make: *** No rule to make target 'clean'.  Stop.
make: *** No targets specified and no makefile found.  Stop.

I am looking through configure to try and find the problem.  I figured I would go ahead and post back here since you have a lot more experience with it than I do.

I don't see a dmd package on Debian.  I found gdmd and installed that.  I am not sure what it is looking for on curl.

$ which curl
/usr/bin/curl
bselme@mj04lv70:~/onedrive$ curl --version
curl 7.67.0 (x86_64-pc-linux-gnu) libcurl/7.67.0 OpenSSL/1.1.1d zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.20.2 (+libidn2/2.0.5) libssh2/1.8.0 nghttp2/1.40.0 librtmp/2.3
Release-Date: 2019-11-06
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets

**Comment by abraunegg** [comment]
@webe3 
> I am running on Debian

Unsure how your initially installing the application, however you need to ensure these dependencies are met: https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#dependencies-ubuntu-18x-ubuntu-19x--debian-9-debian-10---x86_64

**Comment by webe3** [comment]
Debian has a onedrive package so I did not have to build it.  The visit homepage for the package goes to your github page so it is your program.  The version is 2.3.13.

**Comment by abraunegg** [comment]
@webe3 
No problem - @norbusan keeps the Debian packages fairly well up to date.

Most likely your missing `libcurl4-openssl-dev`

Make sure you install these:
```
sudo apt install build-essential
sudo apt install libcurl4-openssl-dev
sudo apt install libsqlite3-dev
sudo apt install pkg-config
sudo apt install git
sudo apt install curl
```

**Comment by webe3** [comment]
The updated client using my userid for the new company onedrive leads me to the same web page saying approval is required for OneDrive client for Linux.  When I use my userid for the OneDrive for the old company that had been working, I get a web page that says onedrive client for linux is unverified.  However, I get the option to cancel or accept.  If I click on accept then it takes me back to the initial screen where I enter my userid and password and then it goes on to where I get the URL to paste into onedrive.
I never had to accept the use of onedrive before which tells me it was approved for everyone.  The fact that the new company onedrive does not even get to the page that asks me to accept means that it is set up differently.  So the solution would be either to have the admins of the new company onedrive or Azure active directory approve the linux onedrive client for everyone or set it up so that third party clients are allowed and each user is prompted to accept.  It seems like those URLs you posted above  show how to do one of the options.  I don't know anything about administering Azure Active Directory and I don't have any permission to do anything anyway.  Hopefully I can get to the right people that can do something.

**Comment by abraunegg** [comment]
@webe3 
> The updated client using my userid for the new company onedrive leads me to the same web page saying approval is required for OneDrive client for Linux. When I use my userid for the OneDrive for the old company that had been working, I get a web page that says onedrive client for linux is unverified.

Correct - because this is a 'new' app .. not setup / configured with the verification part done.

> If I click on accept then it takes me back to the initial screen where I enter my userid and password and then it goes on to where I get the URL to paste into onedrive.

OK .. so you got the correct authentication URL now ?

> I never had to accept the use of onedrive before which tells me it was approved for everyone.

First time accessing a new application id will do this, this is a new application id

> The fact that the new company onedrive does not even get to the page that asks me to accept means that it is set up differently.

This contradicts your statement above

> So the solution would be either to have the admins of the new company onedrive or Azure active directory approve the linux onedrive client for everyone or set it up so that third party clients are allowed and each user is prompted to accept. It seems like those URLs you posted above show how to do one of the options. I don't know anything about administering Azure Active Directory and I don't have any permission to do anything anyway. Hopefully I can get to the right people that can do something.

Correct. This is a Azure Active Directory problem to 'accept' the use of this client. The PR was all about eliminating any doubt around the 'application API access' from needing 'Admin Consent'

Unfortunately I cannot do much else here.

**Comment by webe3** [comment]
No, I do not have the correct URL now.  I still have the userid and pw for onedrive for the old company and a different userid and pw for onedrive for the new company.  I assume that the physical file locations where the files are actually stored on OneDrive are different since there was a migration process to copy from one to the other.  I need to use the new company userid and pw in order to sync with the new company OneDrive.  What I was trying to explain was that using the new version you gave me to build the following occurs:
with the old company userid I got prompted which I accepted and thus got the URL.  I just did this as a test to see if the onedrive client acted any different than it did previously.  However, I assume if I use this URL, it will sync with the old company OneDrive which is no good.

with the new company userid I got the admin approval is required screen so I still do not have a URL to paste into the prompt.

So, I am still waiting for someone to change Azure Active Directory to be able to use the client.

Thanks for your help.  I forwarded the links you posted and tried to explain to the helpdesk that it was an Azure Active Directory problem.
